litgpt/generate/base.py

```diff
@@ -24,13 +24,28 @@ def multinomial_num_samples_1(probs: torch.Tensor) -> torch.Tensor:
     return torch.multinomial(probs, num_samples=1)
 
 
-def sample(logits: torch.Tensor, temperature: float = 1.0, top_k: Optional[int] = None) -> torch.Tensor:
+def sample(
+    logits: torch.Tensor, temperature: float = 1.0, top_k: Optional[int] = None, top_p: Optional[float] = None
+) -> torch.Tensor:
     logits = logits[0, -1]
     # optionally crop the logits to only the top k options
     if top_k is not None:
         v, i = torch.topk(logits, min(top_k, logits.size(-1)))
         # do not use `torch.where` as in nanogpt because it will repeat top-k collisions
         logits = torch.full_like(logits, float("-inf")).scatter_(-1, i, v)
+    # optionally crop the logits to smallest set of logits with a cumulative probability above top_p
+    if top_p is not None:
```

Comment:
I've always thought as `top_k` vs `top_p` not both together. Is this intentional that a user can use both, or should this raise an error if both are not `None`?

Because in theory it would be redundant if your `top_k` selects more than your `top_p` setting, right? I suppose it could be useful to have `tok_k` as a hard cut-off though if `top_p` is too generous, but in that case the if statement for `top_k` should be after the one for `top_p`?

Reply:
Yes it's intentional: i was thinking that one can decide to first select the most top-k probable tokens and then refine the selection dynamically with top-p, which is helpful since after the top-k selection the probability distribution of the next token changes with the top-p acting on the "new" distribution. I also agree that sometimes the combination of the two could return the same distribution but i think it happens for high value of top-p. 

I was also checking the[ HF default and the apply](https://github.com/huggingface/transformers/blob/9fe3f585bb4ea29f209dc705d269fbe292e1128f/src/transformers/generation/utils.py#L735):

1. Temperature scaling
2. Top-k
3. Top-p

Reply:
> most top-k probable tokens and then refine the selection dynamically with top-p, which is helpful since after the top-k selection the probability distribution of the next token changes with the top-p acting on the "new" distribution.

Good point, I didn't think about the rescaling after the top_k.